Refine the tools page side panel

mattsmida opened the pull request:
Implement auto-complete comboboxes for languages and authors on the tools page's side panel.

Closes #700.

lauft commented on themes/bootstrap/layouts/_default/tools.html:
```diff
@@ -53,50 +53,32 @@
         </div>
         <div class="well">
           <fieldset>
```
Let's merge the panels! 🤔

lauft commented on themes/bootstrap/layouts/_default/tools.html:
```diff
@@ -53,50 +53,32 @@
+              <input type="text" name="language-input" id="languages-combo" placeholder="" role="combobox" class="combo-input" autocomplete="off"/>
+              <div class="combo-menu" role="listbox" id="languages-listbox"></div>
             </div>
           </fieldset>
         </div>
         <div class="well">
           <fieldset>
```
Let's merge the panels! 🤔

lauft commented on themes/bootstrap/static/css/multiselect.css:
```diff
@@ -0,0 +1,151 @@
+}
+
+.combo-input:focus {
+  border-color: #e60023;
```
Let's use a more blueish color for a focused combo box.

lauft commented on themes/bootstrap/layouts/_default/tools.html:
```diff
@@ -53,50 +53,32 @@
-          <span data-bind="visible: filteredTools().length == 1">tool/extension</span>
-          matching your search:
+        <p class="lead" id="search-result-message">
+          <!-- There are N tools/extensions matching your search: -->
```
Out of curiosity: `There <span id="search-result-message"> matching your search:` is not possible?

mattsmida replied:
Decided to initialize this with a `loading` message, and then update the whole message on state change. Reason: with the proposed `<span>`, the message's length changes in a jarring way on search.

lauft commented on themes/bootstrap/static/js/filter.js:
```diff
@@ -0,0 +1,228 @@
+  return sortedTools;
+}
+
+/** Given the tools data, return a sorted array of the languages
+ *  to be used for filtering.
+ */
```
Use [Google JavaScript code style](https://google.github.io/styleguide/jsguide.html#jsdoc-general-form) 
```suggestion
/** 
 * Given the tools data, return a sorted array of the languages to be used for filtering.
 */
```
In general, I would format all JSDoc like this (i.e. also those in line 16, 29, ...) and rather use _one sentence, one line_ than _hard wrap at X chars_...

lauft commented on themes/bootstrap/static/js/multiselect.js:
```diff
@@ -0,0 +1,328 @@
+  Type: 9 };
+
+
+// filter an array of options against an input string
+// returns an array of options that begin with the filter string, case-independent
```
Use JSDoc comment for function descriptions.

lauft reviewed (dismissed):
Overall: I like the change, the search bar is much more compact now. 👍🏻 

I have left some comments with some ideas to further improve the layout.

lauft commented on themes/bootstrap/static/css/multiselect.css:
```diff
@@ -0,0 +1,151 @@
+/** Multiselect component adapted from
+    https://webdevtrick.com/html-combobox-with-javascript/
+  */
+
```
The combo box seems to have the same height as the search box. Would it be possible to make them a bit narrower? Could we display a placeholder text like `select language` (or just `select...`, if there is not enough space)?

lauft commented on themes/bootstrap/static/js/filter.js:
```diff
@@ -0,0 +1,228 @@
+            <em class="bi bi-archive"></em>
+            <span class="tooltip-text">project has been archived</span>
+          </span><br>
+          <span style="display: ${tool.owner.length > 0 ? 'visible' : 'none'}>
```
Add missing double colon:
```suggestion
          <span style="display: ${tool.owner.length > 0 ? 'visible' : 'none'}">
```

lauft commented on themes/bootstrap/static/js/filter.js:
```diff
@@ -0,0 +1,228 @@
+        </p>
+        <p>
+          <span>Rating:
+            <span title="Rating">${tool.rating}</span>
+            &ensp;<em class="bi bi-star"></em>
```
Remove linebreak to reduce whitespace between rating and star icon
```suggestion
            <span title="Rating">${tool.rating}</span>&ensp;<em class="bi bi-star"></em>
```

mattsmida commented:
Still pending: filters for category.

mattsmida commented:
In addition to addressing the comments in this thread, this PR now includes a tools page that can be filtered based on category. For ti.net, such filtering is turned off by default -- the plan is to deploy it on the timewarrior site with minimal changes.

mattsmida commented:
I have converted this PR to a draft pending the resolution of [the similar issue on timewarrior.org](https://github.com/GothenburgBitFactory/tw.org/pull/866). Let's finish that one before returning to this PR.
